src/main/java/com/artozersky/HackerNewsAPI/service/impl/NewsPostServiceImpl.java

```diff
@@ -0,0 +1,262 @@
+        NewsPostModelImpl post = postRepository.findById(id)
+        .orElseThrow(() -> new CustomNotFoundException("Post not found with id: " + id));
+        post.upVote();        
+        NewsPostModelImpl updatedPost = postRepository.save(post);
+
+        //if in the cache update the cache
+        if (cacheService.getPostFromCacheById(id) != null) {
+            cacheService.putPostInCache(updatedPost);
+        }
+        //updateCacheWithTopPosts();
+
+
+        NewsPostsResponseDTOImpl responseDTO = modelMapper.map(updatedPost, NewsPostsResponseDTOImpl.class);
+
+        responseDTO.setMessage("Vote updated successfully");
+        
+        return responseDTO;
+    }
+    
+    @Override
+    public NewsPostsResponseDTOImpl downVote(Long id) {
+        NewsPostModelImpl post = postRepository.findById(id)
+        .orElseThrow(() -> new CustomNotFoundException("Post not found with id: " + id));
+        post.downVote();        
+        NewsPostModelImpl updatedPost = postRepository.save(post);
+
+        //downvote can make a post go down in the rank and out of sorted cache
+        updateCacheWithTopPosts();
+
+        NewsPostsResponseDTOImpl responseDTO = modelMapper.map(updatedPost, NewsPostsResponseDTOImpl.class);
+        responseDTO.setMessage("Vote updated successfully");
+        
+        return responseDTO;
+    }
+    
+    @Override
+    public NewsPostsResponseDTOImpl deletePost(Long postId) {
+        
+        NewsPostModelImpl post = postRepository.findById(postId)
+                .orElseThrow(() -> new CustomNotFoundException("Post not found with id: " + postId));
+        
+        postRepository.delete(post);
+
+        updateCacheWithTopPosts();
+
+        NewsPostsResponseDTOImpl responseDTO = new NewsPostsResponseDTOImpl();
+        responseDTO.setPostId(postId);
+        responseDTO.setMessage("Post deleted successfully");
+    
+        return responseDTO;
+    }
+
+    private NewsPostsResponseDTOImpl convertToDTO(NewsPostModelImpl postModel) {
+        NewsPostsResponseDTOImpl responseDTO = modelMapper.map(postModel, NewsPostsResponseDTOImpl.class);
+        responseDTO.setMessage("Successfully fetched " + postModel.getPostId());
+        return responseDTO;
+    }
+
+    public List<NewsPostModelImpl> fetchPostsFromDbIfNeeded(
+    int postsNeeded, 
+    List<NewsPostModelImpl> cachedPosts) {
+
+    if (postsNeeded > 0) {
+
+        Pageable pageable = PageRequest.of(0, postsNeeded);
+
+        List<NewsPostModelImpl> dbPosts = postRepository.findTopPostsByScoreExcludingIds(
+            cachedPosts.stream()
+                .map(NewsPostModelImpl::getPostId)
+                .collect(Collectors.toList()), 
+            pageable);
+
+            //updating this field each time called to db
+            dbPosts.forEach(NewsPostModelImpl::updateElapsedTime);
+            return dbPosts;
+    }
+    
+    return new ArrayList<>();
+}
+
+public void updateCacheWithTopPosts() {
```

Comment:
I think i get what you did here, every time you need to insert into the cache / delete you clear all cache and then populate it from the database again and again every time. i think we need to think of something more elegant, in load this approach will load the system and the database to get all top posts every time you do a small change in the database. also this function does not belong here it needs to be in the cache service.

Reply:
fixed